src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -261,22 +265,33 @@ impl NonOverlapSubLevelPicker {
                 break;
             }
 
+            overlap_len_and_begins.pop();
+
             // check reverse overlap
-            for reverse_index in (0..last_level_index).rev() {
+            for (reverse_index, (old_overlap_len, old_overlap_begin)) in (0..last_level_index)
+                .rev()
+                .zip_eq_fast(overlap_len_and_begins.into_iter().rev())
+            {
                 let target_tables = &l0[reverse_index].table_infos;
-                let overlap_files = overlap_info.check_multiple_overlap(target_tables);
+                let (overlap_files, new_overlap_begin) =
+                    overlap_info.check_multiple_overlap(target_tables);
                 let mut extra_overlap_sst = Vec::with_capacity(overlap_files.len());
-                for other in overlap_files {
+                for (new_overlap_offset, other) in overlap_files.into_iter().enumerate() {
```

Comment:
Great, Do we still need sorting if we expand by offset? https://github.com/risingwavelabs/risingwave/blob/91a6bcc6ca139c908b99961d52b7a1d70a2bdbd2/src/meta/ src/hummock/compaction/picker/min_overlap_compaction_picker.rs#L325

Reply:
> Great, Do we still need sorting if we expand by offset? https://github.com/risingwavelabs/risingwave/blob/91a6bcc6ca139c908b99961d52b7a1d70a2bdbd2/src/meta/ src/hummock/compaction/picker/min_overlap_compaction_picker.rs#L325

This PR does not change corresponding logic on this.